pkg/plugins/plugins.go

```diff
@@ -0,0 +1,63 @@
+package plugins
+
+// OpenShiftAzureConfig is a new, Red Hat owned type defined in an external project.  It
+// represents a superset of the configuration required for all steps in the plugin process *after*
+// the manifest is validated.
+//
+// OpenShiftAzureConfig is a versioned type.  It is expected that the configuration will change
+// between releases and may require migration steps in the future.  The storage of this type
+// is opaque to the plugins, they only require that it be passed in.  However, it must be retained
+// to allow it to be lifecycled from version to version.
+type OpenShiftAzureConfig struct{}
+
+// ManagedOpenShiftCluster is the external schema that is passed to the RP.  It is defined in
+// acs-engine.  ManagedOpenShiftCluster will reuse many of the lower level objects shared by
+// ManagedCluster but is separated in order to keep the relationship with AKS clear.
+type ManagedOpenShiftCluster struct{}
+
+// Validator interface is the first plugin in the chain.  It is responsible for validation of
+// both ManagedOpenShiftCluster.  ManagedOpenShiftCluster enters this process in a versioned state.
+// It is validated both according to the version and against the previous state of the world.
+// Versioned validation should occur first.  The manifest, if acceptable, may then be converted
+// into the internal representation and validated against the previous state of the world.
+//
+// Since this plugin validates with both a versioned manifest and an unversioned manifest it will
+// be necessary to cycle the plugin along side the api versions.
+//
+// TODO: is the lifecycle here acceptable?  It is how api.validate.go works.  Does this make it harder
+// on the RP calling the method?  Might have to provide an api version aware factory to return the correct
+// implementation.
+type Validator interface {
+	Validate(new *ManagedOpenShiftCluster, old *ManagedOpenShiftCluster) (ManagedOpenShiftCluster, []error)
+}
+
+// ConfigManager is responsible for the generate and/or upgrade of the OpenShiftAzureConfig.  The config
+// generated here is a superset of all configurations used by other plugins. If an existing config
+// exists for the ManagedOpenShiftCluster it is passed in so it may be used to transfer values to the new config.
+// A new configuration is always returned.
+type ConfigManager interface {
+	Generate(apimodel *ManagedOpenShiftCluster, existingConfig *OpenShiftAzureConfig) (OpenShiftAzureConfig, error)
+}
+
+// HCPManager generates the values required by helm charts that are created for the customer control plane.
```

Comment:
Does generate also means apply in the cluster or is apply a separate step that needs to be exposed in the interface?

Reply:
I expect that to be a separate step.  I don't think our plugins should be actually creating anything, we should leave that up to the RP since it may be done in parallel or with other interesting requirements.  I do wonder if we might want some sort of lifecycle hook in here that could get access to the cluster if necessary or if that would be a no-no.